FuncTracer.cpp

```diff
@@ -9,26 +9,28 @@
 static std::set<std::string> logged_functions;
 static std::mutex log_mutex;
 
-VOID log_function_call(const char* img_name, const char* func_name)
+void log_function_call(const char* img_name, const char* func_name)
 {
-    // Check if the function is already logged to avoid duplicate logging
-    std::string key;
-    key.append(img_name).append(1, ':').append(func_name);
+    std::string log_key;
     {
         std::lock_guard<std::mutex> guard(log_mutex);
-        if (logged_functions.contains(key))
-            return; // Already logged, skip
-        logged_functions.insert(key);
+        log_key = std::string(img_name) + ":" + func_name;
+        if (logged_functions.contains(log_key))
+            return;
+        logged_functions.insert(log_key);
     }
 
-    std::stringstream ss;
-    PIN_LockClient();
-    pid_t pid = PIN_GetPid();
-    PIN_UnlockClient();
-    ss << "[PID:" << pid << "] [Image:" << img_name << "] [Called:" << func_name << "]\n";
-    LOG(ss.str());
-}
+    pid_t pid;
+    {
```

Comment:
I don't understand the reason for these brackets ?

Reply:
This limits the scope of the lock and unlock operations by grouping them together. It's just good practice, but not strictly necessary at the moment.

Reply:
Then again, I am not an expert in c++, it's just what I've been taught back in school :D

Reply:
Yeah, I see the intent, but at this point, it's just aesthetics since in the scope, there is no new object creation/destruction , and the compiler is going to elide them. So I guess a pair of comment lines would be the same thing (only easier to read)

Reply:
it's not the same; the above brackets contain initialization of 
```C++
 std::lock_guard<std::mutex> guard(log_mutex);
``` 
which is being destroyed at the scope ends, on line 21, thus releasing the mutex lock.

on 
```c++
    pid_t pid;
    {
        PIN_LockClient();
        pid = PIN_GetPid();
        PIN_UnlockClient();
    }
```

instead there is no new object initialization inside brackets, so the scope is useless (and being removed by the compiler. It's even a bit misleading because at a first sight, one can think there is a new scope and need extra reasoning to realize the brackets are useless :)

Reply:
Ah, I see, makes sense. I've removed the redundant scope. Should be all good now.